### PR TITLE
libc/arch_atomic: add gcc legacy __sync buitins support and add FAR to pointers.

### PR DESCRIPTION
**Pull request by CV-Bowen**
## Summary
1. Add gcc legacy __sync builtins support, https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html#g_t_005f_005fsync-Builtins;
2. Add FAR to pointers;

## Impact
all chip select LIBC_ARCH_ATOMIC.

## Testing
CI

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

File: libs/libc/machine/arch_atomic.c [hunk 1]
```diff
@@ -34,7 +34,8 @@
 
 #define STORE(n, type)                                    \
                                                           \
-  void __atomic_store_ ## n (volatile void *ptr,          \
+  void weak_function __atomic_store_ ## n (               \
+                             volatile void *ptr,          \
```
Comment (pkarashchenko) [inline libs/libc/machine/arch_atomic.c:38]: Can we add `FAR` to pointers? Or maybe in the next PR?

File: libs/libc/machine/arch_atomic.c [hunk 2]
```diff
@@ -401,3 +425,35 @@ FETCH_XOR(4, uint32_t)
...
+ * redefined/redeclared problem.
+ */
+
+#ifndef __clang__
+
+/****************************************************************************
+ * Name: __sync_add_and_fetch_1
+ ****************************************************************************/
+
+SYNC_ADD_FETCH(1, uint8_t)
```
Comment (pkarashchenko) [inline libs/libc/machine/arch_atomic.c:439]: do we need `__sync_sub_and_fetch` and others?

Comment (CV-Bowen) [reply libs/libc/machine/arch_atomic.c:439]: Yes, i will do this in this PR later today.

Comment (CV-Bowen) [reply libs/libc/machine/arch_atomic.c:555]: @pkarashchenko Done

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (CV-Bowen) [reply libs/libc/machine/arch_atomic.c]: @xiaoxiang781216 Done

**Comment by pkarashchenko** [comment]
In the link that you posted in PR description I also see `__sync_fetch_and_add` and others. Do we need it?

**Comment by CV-Bowen** [comment]
> In the link that you posted in PR description I also see `__sync_fetch_and_add` and others. Do we need it?

Yes, __atomic_add_fetch and others are also needed and i want to put these modifications to next PR. Actually, the implementation of __atomic_ops_fetch  and __sync_ops_and_fetch is the same and i want to unity them.

**Comment by CV-Bowen** [comment]
> it seems pointless to make those functions `weak`.

Yes, `weak_function` seems unnecessary and i removed them just recently. Please have a look.

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**